sapi/fpm/fpm/fpm_main.c

```diff
@@ -1118,8 +1119,19 @@ static void init_request_info(void)
 				char *ptr;
 
 				if (pt) {
+					// If DocumentRoot contains cyrillic characters and PHP is invoked with SetHandler (not applicable to ProxyPassMatch),
+					// then the cyrillic characters are urlencoded by apache, and we need to decode them, for example with
+					// DocumentRoot /home/hans/web/cyrillicрф.ratma.net/public_html
+					// env_script_filename contains /home/hans/web/cyrillic%D1%80%D1%84.ratma.net/public_html/index.php.
+					// and we must decode it to /home/hans/web/cyrillicрф.ratma.net/public_html/index.php.
+					if(apache_was_here && memchr(pt, '%', len)) {
+						len = php_raw_url_decode(pt, len);
+						ptr = &pt[len]; // php_raw_url_decode() writes a trailing null byte, &pt[len] is that null byte.
+						goto apache_cyrillic_jump;
```

Comment:
This is quite hacky and not sure if it's going to work for another loop part. You should maybe alloc new space and replace pt in this case.

Reply:
>This is quite hacky

yeah well, i couldn't really find a better way without jump🤔 

for example we could do
```c
bool skip_first=false;
if(apache_was_here && memchr(pt, '%', len)) {
len = php_raw_url_decode(pt, len);
ptr = &pt[len]; // php_raw_url_decode() writes a trailing null byte, &pt[len] is that null byte.
skip_first=true;
}
while (skip_first || (ptr = strrchr(pt, '/')) || (ptr = strrchr(pt, '\\'))){
skip_first=false;
```
- now we check a flag with every iteration, and we disable a flag with every iteration, a flag we don't need at all if we use jmp

or we could do
```c
if(apache_was_here && memchr(pt, '%', len)) {
len = php_raw_url_decode(pt, len);
ptr = &pt[len]; // php_raw_url_decode() writes a trailing null byte, &pt[len] is that null byte.
} else {
     (ptr = strrchr(pt, '/')) || (ptr = strrchr(pt, '\\'))
}
do{...} 
while((ptr = strrchr(pt, '/')) || (ptr = strrchr(pt, '\\'));
```
now we avoid the goto and the overhead-per-iteration and the flag but now we have to duplicate the `(ptr = strrchr(pt, '/')) || (ptr = strrchr(pt, '\\'))`  logic

>not sure if it's going to work for another loop part.

i don't understand what you meant, what may not work?

> You should maybe alloc new space and replace pt in this case.

not needed, php_raw_url_decode will never lengthen the string. it may shrink the string, or leave it with the same length, but never grow it.

Reply:
sorry it's fine